src/analysis/processing/qgsalgorithmrasterrank.cpp

```diff
@@ -0,0 +1,309 @@
+{
+  for ( QgsMapLayer *layer : std::as_const( mLayers ) )
+  {
+    layer->moveToThread( QThread::currentThread() );
+  }
+
+  QgsCoordinateReferenceSystem outputCrs;
+  if ( parameters.value( QStringLiteral( "CRS" ) ).isValid() )
+  {
+    outputCrs = parameterAsCrs( parameters, QStringLiteral( "CRS" ), context );
+  }
+  else
+  {
+    outputCrs = mLayers.at( 0 )->crs();
+  }
+
+
+  const Qgis::DataType outputDataType = qobject_cast<QgsRasterLayer *>( mLayers.at( 0 ) )->dataProvider()->dataType( 1 );
+  double outputNoData = 0.0;
+  if ( qobject_cast<QgsRasterLayer *>( mLayers.at( 0 ) )->dataProvider()->sourceHasNoDataValue( 1 ) )
+  {
+    outputNoData = qobject_cast<QgsRasterLayer *>( mLayers.at( 0 ) )->dataProvider()->sourceNoDataValue( 1 );
+  }
+  else
+  {
+    outputNoData = -FLT_MAX;
+  }
+  const bool outputNoDataOverride = parameterAsInt( parameters, QStringLiteral( "NODATA_HANDLING" ), context ) == 1;
+
+  QgsRectangle outputExtent;
+  if ( parameters.value( QStringLiteral( "EXTENT" ) ).isValid() )
+  {
+    outputExtent = parameterAsExtent( parameters, QStringLiteral( "EXTENT" ), context, outputCrs );
+  }
+  else
+  {
+    outputExtent = QgsProcessingUtils::combineLayerExtents( mLayers, outputCrs, context );
+  }
+
+  double minCellSizeX = 1e9;
+  double minCellSizeY = 1e9;
+  for ( QgsMapLayer *layer : mLayers )
+  {
+    QgsRasterLayer *rLayer = qobject_cast<QgsRasterLayer *>( layer );
+
+    QgsRectangle extent = rLayer->extent();
+    if ( rLayer->crs() != outputCrs )
+    {
+      QgsCoordinateTransform ct( rLayer->crs(), outputCrs, context.transformContext() );
+      extent = ct.transformBoundingBox( extent );
+    }
+
+    minCellSizeX = std::min( minCellSizeX, ( extent.xMaximum() - extent.xMinimum() ) / rLayer->width() );
+    minCellSizeY = std::min( minCellSizeY, ( extent.yMaximum() - extent.yMinimum() ) / rLayer->height() );
+  }
+
+  double outputCellSizeX = parameterAsDouble( parameters, QStringLiteral( "CELL_SIZE" ), context );
+  double outputCellSizeY = outputCellSizeX;
+  if ( outputCellSizeX == 0 )
+  {
+    outputCellSizeX = minCellSizeX;
+    outputCellSizeY = minCellSizeY;
+  }
+
+  double cols = std::round( ( outputExtent.xMaximum() - outputExtent.xMinimum() ) / outputCellSizeX );
+  double rows = std::round( ( outputExtent.yMaximum() - outputExtent.yMinimum() ) / outputCellSizeY );
+
+  const QString outputFile = parameterAsOutputLayer( parameters, QStringLiteral( "OUTPUT" ), context );
+  const QFileInfo fi( outputFile );
+  const QString outputFormat = QgsRasterFileWriter::driverForExtension( fi.suffix() );
+
+  auto writer = std::make_unique<QgsRasterFileWriter>( outputFile );
+  writer->setOutputFormat( outputFormat );
+  std::unique_ptr<QgsRasterDataProvider> provider( writer->createMultiBandRaster( outputDataType, cols, rows, outputExtent, outputCrs, mRanks.size() ) );
+  if ( !provider )
+    throw QgsProcessingException( QObject::tr( "Could not create raster output: %1" ).arg( outputFile ) );
```

Comment:
mLayers will be leaked in this scenario -- can you use std::vector< unique_ptr< for mLayers to avoid that?